### PR TITLE
VACMS-11505: add Last Saved by an Editor field to all  content types

### LINKED ISSUE #11505: Add last touched by a human datetime field to all content types
## Description
We need to add a new datetime field to all content types. It will be automagically populated by functionality attached to the save button. This will allow us to track the last time a human reviewed/touched a node. 

## Acceptance Criteria
- [ ] Add datetime field to all content types. The field will be disabled but visible on the form and view in the CMS. 
- [ ] Update the Save button functionality to automatically populate the field with the date and time the node was saved. 

field to be called: Last Saved by an Editor
Helper text: This is the last time an editor interacted with this content. This field is populated automatically when you click Save. 

### CMS Team
Please check the team(s) that will do this work.

- [ ] `Program`
- [ ] `Platform CMS Team`
- [ ] `Sitewide Crew`
- [x] `⭐️ Sitewide CMS`
- [ ] `⭐️ Public Websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`

### PR DESCRIPTION
**Pull request by edmund-dunn**
## Description

Closes #11505.

## Testing done
Tested locally. 

## Screenshots
CleanShot 2022-11-15 at 08 10 04
CleanShot 2022-11-15 at 08 07 53

## QA steps

Select a piece of content, scroll to the bottom and change the moderation state to Published and click save. In the node view, you can see the field at the bottom of the right sidebar with a date and time. If you go back into the edit form, you will be able to see the information in the top section of the right sidebar. 

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `Platform CMS Team`
- [ ] `Sitewide program`
- [x] `⭐️ Sitewide CMS`
- [ ] `⭐️ Public websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`

### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping the UX writer so they are ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping the UX writer to write and queue content
- [ ] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

### CODE REVIEW

File: config/sync/core.entity_form_display.node.health_care_local_facility.default.yml [hunk 1]
```diff
@@ -55,7 +56,6 @@ third_party_settings:
     group_editorial_workflow:
       children:
         - moderation_state
-        - revision_log
```
Comment (swirtSJW) [inline config/sync/core.entity_form_display.node.health_care_local_facility.default.yml:58]: This is AWOL.    Needs to be added back by hand.

File: docroot/modules/custom/va_gov_backend/src/EventSubscriber/LastEditorSaveEventSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  /**
+   * Custom form submit to set the value for the last saved by editor field.
+   *
+   * @param array $form
+   *   The form array.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The form state.
+   */
+  public function lastSavedByEditorSetTimestamp(array $form, FormStateInterface $form_state) {
+    $node = $form_state->getFormObject()->getEntity();
```
Comment (ndouglas) [inline docroot/modules/custom/va_gov_backend/src/EventSubscriber/LastEditorSaveEventSubscriber.php:52]: Just needs a type hint to get rid of the PHPStan gripe.
Screen Shot 2022-11-10 at 7 16 39 AM

File: docroot/modules/custom/va_gov_backend/src/EventSubscriber/LastEditorSaveEventSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+  /**
+   * Form alter Event call.
+   *
+   * @param \Drupal\core_event_dispatcher\Event\Form\FormAlterEvent $event
+   *   The event.
+   */
+  public function formAlter(FormAlterEvent $event): void {
+    $form = &$event->getForm();
+    $form_state = $event->getFormState();
```
Comment (ndouglas) [inline docroot/modules/custom/va_gov_backend/src/EventSubscriber/LastEditorSaveEventSubscriber.php:23]: Do you need `$form_state` for anything?

**Review by ndouglas — CHANGES_REQUESTED** [review]
Looks great, just a couple quibbles from the checks.

**Comment by EWashb** [comment]
All functionality looks/acts as expected in my QA

**Review by EWashb — DISMISSED** [review]
all looks/functions as expected in the QA steps.

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module [hunk 4]
```diff
@@ -1003,6 +1002,8 @@ function va_gov_backend_preprocess_page(&$variables) {
       'vet_center_mobile_vet_center',
       'vet_center_outstation',
     ];
+    $last_saved_by_an_editor = $node->get('field_last_saved_by_an_editor')->value;
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_backend/va_gov_backend.module:1005]: When the field is empty, it is just picking up the current date time.  This needs to be defensive about picking up a non value.
image

image

**Review by swirtSJW — CHANGES_REQUESTED** [review]
This is not working as expected.  When empty, it is displaying current time and date 

Other issues, summary from individual comments:

1. We have a bunch of unexprected changes coming through in the config, some of which were not there the on your first config exports.

   - lazy loading - I think that one is related to a dependabot bump of blazy, so it is unexpected but I thnk innocuouse.   - neds a QA step to double check one of the affected content types just to be sure.
   - field_administration going missing from hiddden.. that one seeems odd and should definitely be double checked for the impact.  
   - loss of meta title from groupping,  should be doublle checked 
    - There is a very wierd addition about time limit.   This one definetly needs checking out 
 2. looose check for string node is casting too wide a net.  See suggested code.
  3. Auto formatting of behat spec tests is making it look like every line changed.  Set your IDE to not format .feature files.  We want the raw output from spec to persist so that we can see only changes to the line that changed.
  4. QA steps are outdated please adjust them to the changes made
  5. Add a QA step to cover checking the dates on nodes that have not been editied sincee this change.

File: config/sync/core.entity_form_display.node.health_care_local_facility.default.yml (hunk 1 above)
Comment (swirtSJW) [reply config/sync/core.entity_form_display.node.health_care_local_facility.default.yml:58]: This has to be put back  or the test will never pass.

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module (hunk 4 above)
Comment (edmund-dunn) [reply docroot/modules/custom/va_gov_backend/va_gov_backend.module:1005]: @swirtSJW explain further

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module [hunk 5]
```diff
@@ -1003,6 +1002,8 @@ function va_gov_backend_preprocess_page(&$variables) {
...
+    $variables['page']['last_saved_by_an_editor'] = $last_saved_by_an_editor ? \Drupal::service('date.formatter')->format($last_saved_by_an_editor, 'custom', 'F j Y g:ia') : 'This needs to be reviewed by an editor.';
```
[the remark this answers is not in the document]
Comment (edmund-dunn) [reply docroot/modules/custom/va_gov_backend/va_gov_backend.module:1006]: @ndouglas no, it is the format used in the recent changes sidebar.

File: composer.json [hunk 6]
```diff
@@ -200,7 +200,7 @@
         "symfony/process": "^4.0",
         "symfony/routing": "^4.0",
         "traviscarden/behat-table-comparison": "~0.2",
-        "va-gov/content-build": "^0.0.3136",
+        "va-gov/content-build": "^0.0.3138",
```
[the remark this answers is not in the document]
Comment (edmund-dunn) [reply composer.json]: I merged in main.

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module (hunk 5 above)
[the remark this answers is not in the document]
Comment (ndouglas) [reply docroot/modules/custom/va_gov_backend/va_gov_backend.module:1006]: Should we formalize it?  Feels weird having it in code like a magic constant.

**Review by laflannery — COMMENTED** [review]
The "Updated By:" field seems a bit misleading.  I believe this is actually the author, the original author, that's what it is in Prod currently.  However it seems like the label was updated here but not the functionality.  For example, I updated a piece of content and the new human field has the correct date and time but it still says updated by CMS Migrator:
image

**Review by swirtSJW — DISMISSED** [review]

**Comment by swirtSJW** [comment]
This message seems a little misleading
image

It was saved by an editor, just not since the field was added.  I realize some of this will go away when we bulk process the published and draft nodes to update that field, but until that happens, AND for the archived nodes that we are not going to touch, this message is not speaking the truth, which leads to confusion and cognition errors.   @BlakeOrgan @EWashb

**Comment by EWashb** [comment]
> This message seems a little misleading image
> 
> It was saved by an editor, just not since the field was added. I realize some of this will go away when we bulk process the published and draft nodes to update that field, but until that happens, AND for the archived nodes that we are not going to touch, this message is not speaking the truth, which leads to confusion and cognition errors. @BlakeOrgan @EWashb
> 
> Suggestions:
> 
> * "Unknown"
> * "Uncertain"
> * "?"
> * "Not recorded"
> * (leave it empty)
> * (hide the display )

But the editor was a human editor in this example? I thought that user was a bot. Maybe we say something to signify that the last save was a bot.

**Comment by edmund-dunn** [comment]
@EWashb in this case it because it hasn't been saved since we added the field.

**Comment by EWashb** [comment]
That's exactly what I was thinking. Maybe we do a slight change in wording here, but possibly being pressed for time makes me wonder if we just hide it for the moment so that we can get an accurate report done before the sprint is over.

**Comment by edmund-dunn** [comment]
I think just put "Unknown" since next sprint it will mostly go away with the script to update the field.

**Comment by BlakeOrgan** [comment]
I think using "Unknown" is the best option. Once that's updated, I think that this would be good to approve from my end.

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module (hunk 4 above)
Comment (edmund-dunn) [reply docroot/modules/custom/va_gov_backend/va_gov_backend.module:1005]: @swirtSJW ok, now I got it. This has been addressed. I added a ternary operator.

**Comment by edmund-dunn** [comment]
> The "Updated By:" field seems a bit misleading. I believe this is actually the author, the original author, that's what it is in Prod currently. However it seems like the label was updated here but not the functionality. For example, I updated a piece of content and the new human field has the correct date and time but it still says updated by CMS Migrator: image

I changed this to be the latest revision author after conferring with @EWashb

**Review by laflannery — APPROVED** [review]
Looks good!

**Comment by swirtSJW** [comment]
I am a bit concerned that this got merged without an passing approval from an engineer.
image

Especially when the last requested change were engineering related.

**Pull request merged**